pennylane/vqe/vqe.py

```diff
@@ -73,54 +79,65 @@ class Hamiltonian:
+    other observables is possible with tensor-valued coefficients, i.e.:
 
-        >>> qml.Hamiltonian([1, 2], [qml.PauliX(0), qml.PauliZ(0) @ qml.PauliZ(1)])
+    >>> H1 = qml.Hamiltonian(torch.tensor([1.]), [qml.PauliX(0)])
+    >>> H2 = qml.Hamiltonian(torch.tensor([2., 3.]), [qml.PauliY(0), qml.PauliX(1)])
+    >>> H3 = qml.Hamiltonian(torch.tensor([1., 2., 3.]), [qml.PauliX(0), qml.PauliY(0), qml.PauliX(1)])
+    >>> H3.compare(H1 + H2)
+    True
 
-        When Hamiltonians are defined using arithmetic operations **inside of QNodes**, constituent observables
-        may be queued as operations/an error may be thrown. Thus, Hamiltonians must be defined either outside of QNodes,
-        or inside of QNodes using the conventional method.
+    .. Warning::
+        When Hamiltonians are defined using arithmetic operations **inside of quantum functions**, constituent observables
+        may be queued as operations/an error may be thrown. Thus, Hamiltonians must be defined either outside
+        of quantum functions, or inside of quantum functions using the conventional construction method.
 
-        Note that this issue also arises when calling the ``simplify()`` method.
     """
-    # Todo: this is a temporary solution to make the circuit drawer work
-    num_params = 0
 
-    def __init__(self, coeffs, observables, simplify=False):
+    num_wires = qml.operation.AnyWires
+    num_params = 1
+    par_domain = "A"
+    grad_method = "A"  # supports analytic gradients
 
-        if len(coeffs) != len(observables):
+    def __init__(self, coeffs, observables, simplify=False, id=None, do_queue=True):
+
+        if qml.math.shape(coeffs)[0] != len(observables):
             raise ValueError(
                 "Could not create valid Hamiltonian; "
                 "number of coefficients and operators does not match."
             )
 
-        if any(np.imag(coeffs) != 0):
-            raise ValueError(
-                "Could not create valid Hamiltonian; " "coefficients are not real-valued."
-            )
-
         for obs in observables:
             if not isinstance(obs, Observable):
                 raise ValueError(
                     "Could not create circuits. Some or all observables are not valid."
                 )
 
-        self._coeffs = list(coeffs)
+        self._coeffs = coeffs
         self._ops = list(observables)
+        self._wires = qml.wires.Wires.all_wires([op.wires for op in self.ops], sort=True)
 
-        self.data = []
         self.return_type = None
 
         if simplify:
             self.simplify()
 
-        self.queue()
+        coeffs_flat = [coeffs[i] for i in range(qml.math.shape(coeffs)[0])]
+        # overwrite this attribute, now that we have the correct info
+        self.num_params = len(coeffs_flat)
+        # create an operator using each coefficient as a separate parameter
+        super().__init__(*coeffs_flat, wires=self._wires, id=id, do_queue=do_queue)
```

Comment:
As a summary: `H.coeffs` will be the original tensor, but `H.data` (which is used for tape parameter extraction) will be a list of tensor scalars.

Reply:
This is a very neat solution! Perhaps worth adding your comment above as a code comment?

Reply:
I have a feeling that

```python
coeffs_flat = [coeffs[i] for i in range(qml.math.shape(coeffs)[0])]
```

should be the same as

```python
coeffs_flat = qml.math.unstack(coeffs)
```

But if I recall this didn't work with NumPy?

Reply:
Yes, does not exist for numpy...

Reply:
Silly question, but flattening tensors like this comes up all the time, can we implement our own version of `unstack` in the `qml.math` module, or is that module purely just for sorting out and calling the functions from the different frameworks?